gpflow/optimizers/natgrad.py

```diff
@@ -102,14 +106,10 @@ def minimize(self, loss_fn: Callable, var_list: List[Parameter]):
                 ```
         """
         parameters = [(v[0], v[1], v[2] if len(v) > 2 else XiNat()) for v in var_list]
-        self._natgrad_steps(loss_fn, parameters)
-
-    def _natgrad_steps(self, loss_fn: Callable, parameters: List[Tuple[Parameter, Parameter, XiTransform]]):
-        def natural_gradient_step(q_mu, q_sqrt, xi_transform):
-            self._natgrad_step(loss_fn, q_mu, q_sqrt, xi_transform)
 
         with tf.name_scope(f"{self._name}/natural_gradient_steps"):
-            list(map(natural_gradient_step, *zip(*parameters)))
+            for q_mu, q_sqrt, xi_transform in parameters:
+                self._natgrad_step(loss_fn, q_mu, q_sqrt, xi_transform)
```

Comment:
This was written for `tf.function`. The `autograph=True` might recognise `map` function and apply parallerisable map function instead of sequential `for` loop.

Reply:
here's the optimised code for the original
```
def tf___natgrad_steps(loss_fn: Callable, parameters: List[Tuple[Parameter, Parameter, XiTransform]]):
  with ag__.FunctionScope('_natgrad_steps', 'fscope', ag__.ConversionOptions(recursive=True, user_requested=True, optional_features=(), internal_convert_user_code=True)) as fscope:

    @ag__.do_not_convert_internal
    def natural_gradient_step(q_mu, q_sqrt, xi_transform):
      with ag__.FunctionScope('natural_gradient_step', 'fscope_1', ag__.ConversionOptions(recursive=True, user_requested=True, optional_features=(), internal_convert_user_code=True)) as fscope_1:
        ag__.converted_call(self._natgrad_step, (loss_fn, q_mu, q_sqrt, xi_transform), None, fscope_1)
    with tf.name_scope(f'{self._name}/natural_gradient_steps'):
      ag__.converted_call(list, (ag__.converted_call(map, (natural_gradient_step,) + tuple(ag__.converted_call(zip, tuple(parameters), None, fscope)), None, fscope),), None, fscope)
```
and the new version
```
def tf__f():
  with ag__.FunctionScope('f', 'fscope', ag__.ConversionOptions(recursive=True, user_requested=True, optional_features=(), internal_convert_user_code=True)) as fscope:
    with tf.name_scope(f'{self._name}/natural_gradient_steps'):

      def get_state():
        return ()

      def set_state(_):
        pass

      def loop_body(iterates):
        q_mu, q_sqrt, xi_transform = iterates
        ag__.converted_call(self._natgrad_step, (loss_fn, q_mu, q_sqrt, xi_transform), None, fscope)
        return ()
      ag__.for_stmt(parameters, None, loop_body, get_state, set_state, (), (), ())
```
not sure how to read them. will have a look

Reply:
do we _need_ it to recognise `map`? If so, I propose we use a tensorflow function so that we're sure

Reply:
it looks like, from the [docs](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/control_flow.md)

> Key Point: Only statements that are conditioned on, or iterate over, a TensorFlow object such as tf.Tensor, are converted into TensorFlow ops.

that neither version is likely to be converted to a tf op.

Reply:
but since I can't be sure, I'm just going to put it back how it was for now

Reply:
It looks like you caught an inefficiency in our natgrad code though - could you open an issue for this ?